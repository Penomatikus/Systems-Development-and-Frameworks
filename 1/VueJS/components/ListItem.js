export default Vue.component('list-item', {
    props: ['todo'], 
    data: () => {
        return {
            editMode: false, 
            newMessage: ""
        }
    },
    template: `
        <form id="test">
            <li v-if="editMode == true" >                
                <input type="text" placeholder="enter new value" v-on:input="updateMessage"> 
                <button v-on:click="saveMe(todo)" type="button">
                    Save
                </button>
                <button v-on:click="cancelMe" type="button">
                    Cancel
                </button>
                <button v-on:click="deleteMe(todo)" type="button">
                    Delete
            </button>
            </li>
            <li v-else >                
                {{todo.id}}. {{todo.message}} 
                <button v-on:click="editMe" type="button">
                    Edit
                </button>
                <button v-on:click="deleteMe(todo)" type="button">
                    Delete
                </button>

            </li>
        </form> 
    `,
    methods: {
        editMe: function(event) {
            this.editMode = true;
        },
        
        saveMe: function(todo) {
            // console.log("(Kind)\nHallo todo[" + todo.id + " | " + todo.message + "]" 
            //             + "\". Dank v-on:input=\"updateMessage\" im imput-tag, wurdest hoffentlich geupdated!\n"
            //             + "Meine Eltern werden es nach der n√§chsten Zeile im Code dann bestimmt sehen.")
            this.todo.message = this.newMessage
            this.$emit('update-todo', todo)
            this.editMode = false
        },

        deleteMe: function(todo) {
            this.$emit('delete-todo', todo.id)
        },

        cancelMe: function(event) {
            this.editMode = false;
        }, 

        updateMessage: function(event) {
            this.newMessage = event.target.value
        }
    }
})
