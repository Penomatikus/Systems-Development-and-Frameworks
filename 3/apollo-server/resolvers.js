import GraphQLJSON from "graphql-type-json";

export default {
  JSON: GraphQLJSON,

  Query: {
    hello: (root, { name }) => `Hello ${name || "World"}!`,
    todos: (root, args, { ds }) => ds.getTodos(),
    todo: (root, { id }, { ds }) => { let debugtodo = ds.findTodo(id);console.log(debugtodo);return debugtodo}
  },

  Mutation: {
    myMutation: (root, args, context) => {
      const message = "My mutation completed!";
      context.pubsub.publish("hey", { mySub: message });
      return message;
    },

    addTodo: (root, { id, newMessage }, { ds }) => {
      const todo = {
        // id: new Date().getUTCMilliseconds(),
        id: id,
        message: newMessage
      };

      ds.addTodo(todo);

      return todo;
    },

    updateTodo: (root, { id, updateMessage }, { ds }) => {
      const toBeUpdated = ds.findTodo(id).message; // not save yet!(check undefined)
      ds.updateTodo(id, updateMessage);


      return `todo with ID: ${id} was updated from ${toBeUpdated} to ${updateMessage}`;
    },

    deleteTodo: (root, { id }, { ds }) => {
      ds.deleteTodo(id);
      return `todo with ID ${id} was deleted`;
    }
  }
};
