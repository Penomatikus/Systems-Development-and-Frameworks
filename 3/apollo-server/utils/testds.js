const { DataSource } = require('apollo-datasource');

function searchelement(mok,id){
for (var i=0;i<mok.length;i++) {
  if (mok[i].id == id) 
  {
    return mok[i]
  }
}
return undefined}

export class TodoAPI extends DataSource {
  
  constructor( store ) {
    super();
    this.store = store;
  }

  initialize(config) {
    this.context = config.context;
  }
  
  findTodo(id) {
    return searchelement(mok, id)
  }

  getTodos() {
   return mok
  }

  deleteTodo(id) {
   return mok.splice( mok.indexOf({id: id}), 1 );
  }

  addTodo(todo) {
   return mok.push(todo)
  }

  updateTodo(id, newmessage){
   var tmp = findTodo(id);
   if (tmp == undefined){return undefined}
   tmp.message = newmessage;
   return tmp
  }


}

export const ds = new TodoAPI()

