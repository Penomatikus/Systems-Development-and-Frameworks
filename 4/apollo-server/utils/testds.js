const { DataSource } = require('apollo-datasource');

function searchelement(mok, id){
for (var i=0;i<mok.length;i++) {
  if (mok[i].id == id) 
  {
    return mok[i]
  }
}
return "findTodo: element not found"
}

function filterTodos(mok, userAuth) {
  var retArray = [];

  for (var i=0;i<mok.length;i++) {
    if (mok[i].userAuth == userAuth) 
    {      
      retArray.push(mok[i])
    }
  }

  if(retArray.length == 0)
    console.log("no valid todos found for user")
  return retArray
}

export class TodoAPI extends DataSource {
  
  constructor( store ) {
    super();
    this.store = store;
  }

  initialize(config) {
    this.context = config.context;
  }
  
  findTodo(id, userAuth) {
    return searchelement(this.store, id, userAuth)
  }

  getAllTodos() {
    return this.store;
  }

  getTodosForUser(userAuth) {
    return filterTodos(this.store, userAuth);
  }

  deleteTodo(id) {
   return this.store.splice( this.store.indexOf({id: id}), 1 );
  }

  addTodo(todo, userAuth) {
   return this.store.push({id: todo.id, message: todo.message, userAuth: userAuth})
  }

  updateTodo(id, newmessage){
   var tmp = this.findTodo(id);
   if (tmp == undefined){return undefined}
   tmp.message = newmessage;
   return tmp
  }

}

export const ds = new TodoAPI()

