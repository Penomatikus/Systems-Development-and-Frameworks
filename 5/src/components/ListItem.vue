<template>
    <div id="list-item">
        <form id="test">
            <li v-if="editMode == true">
                <input
                    type="text"
                    placeholder="enter new value"
                    v-on:input="updateMessage"
                />
                <button v-on:click="saveMe(todo)" type="button" id="save">
                    Save
                </button>
                <button v-on:click="cancelMe" type="button" id="cancel">
                    Cancel
                </button>
            </li>
            <li v-else>
                {{ todo.id }}. {{ todo.message }}
                <button v-on:click="editMe" type="button" id="edit">
                    Edit
                </button>
                <button v-on:click="deleteMe(todo)" type="button" id="delete">
                    Delete
                </button>
            </li>
        </form>
    </div>
</template>

<script>
export default {
    name: 'list-item',
    props: ['todo'],
    data: () => {
        return {
            editMode: false,
            newMessage: '',
        }
    },
    methods: {
        editMe: function() {
            this.editMode = true
        },
        saveMe: function(todo) {
            this.todo.message = this.newMessage
            this.$emit('update-todo', todo)
            this.editMode = false
        },
        deleteMe: function(todo) {
            this.$emit('delete-todo', todo.id)
        },
        cancelMe: function() {
            this.editMode = false
        },
        updateMessage: function(event) {
            this.newMessage = event.target.value
        },
    },
}
</script>
