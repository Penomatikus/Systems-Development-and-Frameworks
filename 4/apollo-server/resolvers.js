import GraphQLJSON from 'graphql-type-json'
import { decodeJwt } from './utils/jwtCreator'

export default {
    JSON: GraphQLJSON,

    Query: {
        todos: async (root, { FILTER_MODE }, { dataSources }) => {
            let todos = await dataSources.ds.getAllTodos(FILTER_MODE)
            return todos
        },
        todosForUser: (root, { userAuth }, { dataSources }) => {
            if (decodeJwt(userAuth, 'secret')) {
                return dataSources.ds.getTodosForUser(userAuth)
            }
            return [{ message: 'SECRET NOT VALID' }]
        },
        todo: async (root, { id }, { dataSources }) => {
            let debugtodo = await dataSources.ds.findTodo(id)
            return debugtodo
        },
    },

    Mutation: {
        addTodo: async (
            root,
            { id, newMessage, userAuth },
            { dataSources }
        ) => {
            const todo = {
                id: id,
                message: newMessage,
            }

            if (decodeJwt(userAuth, 'secret')) {
                await dataSources.ds.addTodo(todo, userAuth)
                return todo
            }

            return undefined
        },

        updateTodo: async (root, { id, updateMessage }, { dataSources }) => {
            await dataSources.ds.updateTodo(id, updateMessage)

            return `todo with ID: ${id} was updated to ${updateMessage}`
        },

        deleteTodo: (root, { id }, { dataSources }) => {
            dataSources.ds.deleteTodo(id)
            return `todo with ID ${id} was deleted`
        },
    },
}
