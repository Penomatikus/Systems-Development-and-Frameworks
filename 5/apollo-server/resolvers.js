import GraphQLJSON from 'graphql-type-json'
import { decodeJwt } from './utils/jwtCreator'



export default {
    JSON: GraphQLJSON,

    Query: {
        todos: async (root, { FILTER_MODE, skip, limit }, { dataSources }) => {
            let todos = await dataSources.ds.getAllTodos(
                FILTER_MODE,
                skip,
                limit
            )
            return todos
        },
        todosForUser: (root, { userAuth }, { dataSources }) => {
            if (decodeJwt(userAuth, 'secret')) {
                return dataSources.ds.getTodosForUser(userAuth)
            }
            return [{ message: 'SECRET NOT VALID' }]
        },
        todo: async (root, { id }, { dataSources }) => {
            let debugtodo = await dataSources.ds.findTodo(id)
            return debugtodo
        },
        getToDoDependencies: async (
            root,
            { id},
            { dataSources }
        ) => {
            return await dataSources.ds.getToDoDependencies(id)
        },
    },

    Mutation: {
        addTodo: async (
            root,
            { id, newMessage, userAuth, lastEdited },
            { dataSources }
        ) => {
            const todo = {
                id: id,
                message: newMessage,
            }

            if (decodeJwt(userAuth, 'secret')) {
                await dataSources.ds.addTodo(todo, userAuth, lastEdited)
                return todo
            }

            return undefined
        },

        updateTodo: async (root, { id, updateMessage, userAuth, lastEdited }, { dataSources }) => {
            await dataSources.ds.updateTodo(id, updateMessage, lastEdited)

            return `todo with ID: ${id} was updated to ${updateMessage}`
        },

        addToDoDependency: async (
            root,
            { id, dependencyId },
            { dataSources }
        ) => {
            await dataSources.ds.addToDoDependency(id, dependencyId)
        },

        deleteTodo: (root, { id }, { dataSources }) => {
            dataSources.ds.deleteTodo(id)
            return `todo with ID ${id} was deleted`
        },
    },
}
