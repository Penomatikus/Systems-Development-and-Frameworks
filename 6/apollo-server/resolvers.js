import GraphQLJSON from 'graphql-type-json'
import { decodeJwt, createJwt } from './utils/jwtCreator'
import { ds } from './utils/TodoNeo4JAPI'



export default {
    JSON: GraphQLJSON,

    Query: {
        todos: async (root, { FILTER_MODE, skip, limit }, { dataSources }) => {
            let todos = await dataSources.ds.getAllTodos(
                FILTER_MODE,
                skip,
                limit
            )
            return todos
        },
        todosForUser: async (root, { username, token }, { dataSources }) => {
            const user = await dataSources.ds.getUser(username)
            if (decodeJwt(token, user.pw)) {
                return dataSources.ds.getTodosForUser(username)
            }
            return [{ message: 'SECRET NOT VALID' }]
        },
        todo: async (root, { id }, { dataSources }) => {
            let debugtodo = await dataSources.ds.findTodo(id)
            return debugtodo
        },
        getToDoDependencies: async (
            root,
            { id},
            { dataSources }
        ) => {
            return await dataSources.ds.getToDoDependencies(id)
        },
    },

    Mutation: {
        addTodo: async (
            root,
            { id, newMessage, loginData, lastEdited },
            { dataSources }
        ) => {
            const todo = {
                id: id,
                message: newMessage,
            }
            const user = await dataSources.ds.getUser(loginData.username)
            if (decodeJwt(loginData.token, user.pw)) {
                await dataSources.ds.addTodo(todo, loginData.username, lastEdited)
                return todo
            }
            console.log("permission denied")
            return undefined
        },
        authenticate: async (
            root,
            { Credentials },
            { dataSources }
        ) => {
            console.log("Before JWT Token")
            const user = await dataSources.ds.getUser(Credentials.username)
            console.log("JWT user:" + user)
            console.log("JWT pw:" + Credentials.pw)
            if (dataSources.ds.userExists(Credentials.username) == false || user.pw != Credentials.pw){
                console.log("permission denied")
                return ""
            }
            
            const token = createJwt(Credentials.pw)
            console.log("nice Hello")
            return token
         },
        updateTodo: async (root, { id, updateMessage, loginData, lastEdited }, { dataSources }) => {
            console.log("Update Todo")
            await dataSources.ds.updateTodo(id, updateMessage, lastEdited)

            return `todo with ID: ${id} was updated to ${updateMessage}`
        },

        addToDoDependency: async (
            root,
            { id, dependencyId },
            { dataSources }
        ) => {
            await dataSources.ds.addToDoDependency(id, dependencyId)
        },

        deleteTodo: (root, { id }, { dataSources }) => {
            dataSources.ds.deleteTodo(id)
            return `todo with ID ${id} was deleted`
        },
    },
}
