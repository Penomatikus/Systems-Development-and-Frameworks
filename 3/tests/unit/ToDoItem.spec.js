import { shallowMount } from "@vue/test-utils";
import ListItem from "@/components/ListItem.vue";

import fs from 'fs'
import path from 'path'
import resolvers from '../../apollo-server/resolvers'

import { TodoAPI } from '../../apollo-server/utils/testds'

const { ApolloServer, gql } = require('apollo-server');
const { createTestClient } = require('apollo-server-testing');

const typeDefs = fs.readFileSync(path.resolve(__dirname, '../../apollo-server/schema.graphql'), { encoding: 'utf8' })

const mok = [{id: 1,message:"a"},{id: 2,message:"b"}];  
const mocksds = new TodoAPI(mok);



const allMocks = () => mocksds
 
it('fetches single launch', async () => {
  
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({ mocksds }),
    context: () => ({id: 1,message:"a"},{id: 2,message:"b"}),
    mockEntireSchema: false,

});

server.listen().then(({ url }) => {
  console.log(`ðŸš€  Server ready at ${url}`);
});

const { query } = createTestClient(server);
const GET_TODO = gql`
  query todo($id: Int!) {
    todo(id: $id) {
      id
      message
    }
  }
`;

const res = await query(
			{ query: GET_TODO, 
			  variables: { id: 1 } 
			}
);
console.log(res)
expect(res).toMatchSnapshot();


});

