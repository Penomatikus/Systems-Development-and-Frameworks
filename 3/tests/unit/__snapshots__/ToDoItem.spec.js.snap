// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches single launch 1`] = `
Object {
  "data": Object {
    "todo": null,
  },
  "errors": Array [
    [GraphQLError: Cannot read property 'get' of undefined],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
